src/gf.c

```diff
@@ -1039,9 +1056,35 @@ jl_lambda_info_t *jl_get_specialization1(jl_tupletype_t *types)
     return NULL;
 }
 
-JL_DLLEXPORT void jl_compile_hint(jl_tupletype_t *types)
+// Specialize, but return NULL if there is a call ambiguity
+jl_lambda_info_t *jl_checked_specialization1(jl_tupletype_t *types)
```

Comment:
Actually, codegen.cpp currently never emits MethodErrors directly. It just falls through to calling jl_apply_generic, and lets that handle throwing the error. So this change to codegen is totally separate and I believe not needed for this.